### PR TITLE
ta: avb: RSA public key validation

### PR DESCRIPTION
**Pull request by igoropaniuk**
Introduce support for validation of RSA public key, used to verify
vbmeta partition. By validation it means that the given public key
is trusted and can be used for verification of vbmeta image signature.
For additional details check `validate_vbmeta_public_key` in [1].

It is also required by AVB spec [1] that trusted key is stored in the
tamper-evident storage.

The key data provided from NW is serialized in this format:
```
+--------------+----------------------------+
| Name         | Number of bytes            |
+--------------+----------------------------+
| key_num_bits | 4 ("modulus size in bits") |
| n0inv        | 4 ("-1/n[0] (mod 2^32)")   |
| n            | key_num_bits / 8 (modulus) |
| rr           | key_num_bits / 8 (R^2)     |
+--------------+----------------------------+
```

So, for example, for 4096 bit(default one) key representation we need
1032 bytes
(4 + 4 + 512 + 512) to store these values.

Public exponent is assumed to always be 65537.

Link: [1] https://android.googlesource.com/platform/external/avb/+/master/README.md#Named-Persistent-Values

FYI: there is no any function yet to store initial public key data (for example, during factory provisioning), will add it in this commit after I receive initial feedback (I'm not sure that it should be implemented in this way).

`Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>`

### CODE REVIEW

**Comment by jbech-linaro** [comment]
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

File: ta/avb/entry.c [hunk 1]
```diff
@@ -337,6 +338,69 @@ static TEE_Result read_persist_value(uint32_t pt,
...
 
+static TEE_Result validate_public_key(uint32_t pt,
+				      TEE_Param params[TEE_NUM_PARAMS])
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+	uint32_t flags = TEE_DATA_FLAG_ACCESS_READ |
+			 TEE_DATA_FLAG_ACCESS_WRITE;
```
Comment (jenswi-linaro) [inline ta/avb/entry.c]: Is write access really needed?

File: ta/avb/entry.c [hunk 2]
```diff
@@ -337,6 +338,69 @@ static TEE_Result read_persist_value(uint32_t pt,
...
+				       sizeof(pkey_obj_name), flags, &h);
+	if (res) {
+		EMSG("Can't open public key data object, res = 0x%x", res);
+		return res;
+	}
+
+	res = TEE_GetObjectInfo1(h, &obj_info);
+	if (res) {
+		EMSG("Can't get public key data object info, res = 0x%x", res);
+		return res;
```
Comment (jenswi-linaro) [inline ta/avb/entry.c]: Leaking the handle `h`

File: ta/avb/entry.c [hunk 3]
```diff
@@ -337,6 +338,69 @@ static TEE_Result read_persist_value(uint32_t pt,
...
+	if (TEE_MemCompare(pkey_trusted, pkey_to_validate, pkey_trusted_sz)) {
+		EMSG("Public key validation failed, keys are not identical");
+		res = TEE_ERROR_SECURITY;
+	}
+
+
+out:
+	TEE_CloseObject(h);
+
+	return res;
```
Comment (jenswi-linaro) [inline ta/avb/entry.c]: Leaking `pkey_trusted`

**Comment by igoropaniuk** [comment]
@jenswi-linaro 

>How is the public key provisioned so the TA can read it?

>FYI: there is no any function yet to store initial public key data (for example, during factory provisioning), will add it in this commit after I receive initial feedback (I'm not sure that it should be implemented in this way).

I'll add it to this PR if there is no any objection how the key is validated.

**Comment by igoropaniuk** [comment]
@jenswi-linaro @jbech-linaro also additional question here:
should we provide apossibility of replacing existing public key? or writing of key should be done once, during the factory provisioning ?

Thanks

**Comment by jbech-linaro** [comment]
> @jenswi-linaro @jbech-linaro also additional question here:
> should we provide apossibility of replacing existing public key? or writing of key should be done once, during the factory provisioning ?

I would say it depends on the use case. This is an AVB specific thing and not a generic security feature. So I'd ask you to read up on the AVB specification to see whether that is something that is a) required b) recommend or c) up to the user to decide. When we know that we have better base to make a decision whether we should allow that or not and then also discuss how to do it in a secure way.

**Comment by igoropaniuk** [comment]
> > @jenswi-linaro @jbech-linaro also additional question here:
> > should we provide apossibility of replacing existing public key? or writing of key should be done once, during the factory provisioning ?
> 
> I would say it depends on the use case. This is an AVB specific thing and not a generic security feature. So I'd ask you to read up on the AVB specification to see whether that is something that is a) required b) recommend or c) up to the user to decide. When we know that we have better base to make a decision whether we should allow that or not and then also discuss how to do it in a secure way.

I already did, there in no any notices about the initial key provisioning/key updates. In the current implementation public key is embedded in the U-boot binary, and in case if manufacturer wants to replace it (for example, in case of private key leakage etc.), besides the new set of android partitions (including `vbmeta`), he(manufacturer) delivers updated u-boot blob with embedded new public key and that's it.

In the similar case but with TA usage, the operation for updating of public key in the secure storage should be somehow authorized, and it's completely a grey area for me for now.

File: ta/avb/entry.c (hunk 1 above)
Comment (igoropaniuk) [reply ta/avb/entry.c]: no, will fix

**Comment by jenswi-linaro** [comment]
How about allowing writing the public key if none is written yet and the device is still in unlocked state? As a way of allowing updating the public key if the lock state is changed from locked to unlocked the public key it's possible to overwrite the key once until the session is closed.

This would be an extension in the spirit of https://android.googlesource.com/platform/external/avb/+/master/README.md#Locked-and-Unlocked-mode

**Comment by igoropaniuk** [comment]
@jenswi-linaro 

The problem is that should we allow the transition back to the locked state.
A possible the root of trust attack:

"unlock" bootloader -> tamper needed partition -> recalculate hashes and generate new vbmeta signed by attackers key -> write attackers public key to the secure storage (TA accepts it, as we're in "unlocked" state) -> transition to "lock" state (???)

**Comment by jenswi-linaro** [comment]
Yes, this is at the same level of security as the rollback counters.

We could add a config flag to enable replacement of the public key using the scheme I described above. Even if it's not used directly it would still serve as an example of how to replace the public key.

File: ta/avb/entry.c [hunk 4]
```diff
@@ -336,6 +337,107 @@ static TEE_Result read_persist_value(uint32_t pt,
...
+	uint32_t flags = TEE_DATA_FLAG_ACCESS_READ |
+			 TEE_DATA_FLAG_ACCESS_WRITE;
+
+	if (pt != exp_pt)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	char *pkey_to_provision = params[0].memref.buffer;
+	uint32_t pkey_to_provision_sz = params[0].memref.size;
+
+	res = TEE_OpenPersistentObject(storageid, pkey_obj_name,
```
Comment (jenswi-linaro) [inline ta/avb/entry.c]: This is not needed, if the overwrite flag isn't provided `TEE_CreatePersistentObject()` will not overwrite the object.

Comment (igoropaniuk) [reply ta/avb/entry.c]: ok, will remove

Comment (igoropaniuk) [reply ta/avb/entry.c]: removed

**Comment by igoropaniuk** [comment]
btw, my suggestion is to leave the key provisioning function as it is for now and don't permit updates of the key (until it isn't totally clear to us). 

I'll try to reach AVB spec. authors to clarify this.

**Comment by igoropaniuk** [comment]
Ping. Any other comments/suggestions?

**Comment by jbech-linaro** [comment]
Not as such, but didn't we decide to hold off with this until we had spoken to the Google guys about it?

**Comment by igoropaniuk** [comment]
> Not as such, but didn't we decide to hold off with this until we had spoken to the Google guys about it?

we can add support for updating public key in a different PR (this is what I actually meant :) ).

**Comment by igoropaniuk** [comment]
FYI: I've sent an email with all these questions to David Zeuthen (zeuthen@google.com),  who, based on git logs,  is the main contributor to the AVB2.0 README. Let's wait for at least 1 week for feedback.

**Comment by igoropaniuk** [comment]
@jenswi-linaro 
>How about allowing writing the public key if none is written yet and the device is still in unlocked state? As a way of allowing updating the public key if the lock state is changed from locked to unlocked the public key it's possible to overwrite the key once until the session is closed.

This would be an extension in the spirit of https://android.googlesource.com/platform/external/avb/+/master/README.md#Locked-and-Unlocked-mode

I was thinking about your suggestion, and probably it should be vice-versa. We should permit to update the AVB pubkey in Secure Storage only when we're in "locked" mode. Taking into account that in "locked" state partition tampering leads to stopping the boot process, there two conditions: we're booted in "safe" environment or boot fails.  

In this "safe" environment all requests to update (for example, flashing new Android version, where vbmeta is signed with different AVB private key) the AVB pubkey are by definition also "safe"

**Comment by igoropaniuk** [comment]
Sorry @jenswi-linaro,
I just re-read you message twice and seems that I misunderstood you from the beginning :)
You're right

**Pull request closed without merging**